File: core/src/protocol.zig
const std = @import("std");

pub const protocol_version = "1.0";
pub const max_message_bytes: usize = 16 * 1024 * 1024;

pub const Operation = enum {
    protocol_info,
    inspect_source,
    validate_proposal,
    validate_revision,
    assemble_final,
    apply_final,
};

pub const Request = struct {
    protocol_version: []const u8,
    request_id: []const u8,
    operation: Operation,
    payload_schema: []const u8,
    payload: std.json.Value,
};

pub const ErrorDetail = struct { pointer: ?[]const u8 = null, reason: []const u8 };
pub const ProtocolError = struct {
    code: []const u8,
    message: []const u8,
    retryable: bool,
    details: ?[]const ErrorDetail = null,
};
pub const Success = struct {
    protocol_version: []const u8 = protocol_version,
    request_id: []const u8,
    ok: bool = true,
    result_schema: []const u8,
    result: std.json.Value,
};
pub const Failure = struct {
    protocol_version: []const u8 = protocol_version,
    request_id: []const u8,
    ok: bool = false,
    @"error": ProtocolError,
};

pub const ValueType = enum { string, number, integer, boolean, null, array, object };
pub const RedactedValue = struct {
    visibility: enum { redacted },
    secret_ref: []const u8,
    value_type: ValueType,
    fingerprint: []const u8,

    pub fn isValid(self: RedactedValue) bool {
        const prefix = "hmac-sha256:";
        if (self.secret_ref.len == 0 or self.fingerprint.len != prefix.len + 64 or
            !std.mem.startsWith(u8, self.fingerprint, prefix)) return false;
        for (self.fingerprint[prefix.len..]) |c|
            if (!std.ascii.isHex(c) or std.ascii.isUpper(c)) return false;
        return true;
    }
};

pub const ProtocolInfo = struct {
    protocol_versions: []const []const u8,
    operations: []const []const u8,
    schemas: []const []const u8,
    limits: struct { message_bytes: usize },
};

pub const ParsedRequest = std.json.Parsed(Request);

pub fn parseStrict(comptime T: type, allocator: std.mem.Allocator, bytes: []const u8) !std.json.Parsed(T) {
    if (bytes.len > max_message_bytes) return error.MessageTooLarge;
    return std.json.parseFromSlice(T, allocator, bytes, .{}) catch |err| return switch (err) {
        error.DuplicateField => error.DuplicateField,
        error.UnknownField => error.UnknownField,
        else => error.InvalidJson,
    };
}

pub fn parseRequest(allocator: std.mem.Allocator, bytes: []const u8) !ParsedRequest {
    if (bytes.len > max_message_bytes) return error.MessageTooLarge;
    var parsed = try parseStrict(Request, allocator, bytes);
    errdefer parsed.deinit();
    const request = parsed.value;
    if (!std.mem.eql(u8, request.protocol_version, protocol_version))
        return error.UnsupportedProtocolVersion;
    if (request.request_id.len == 0 or request.request_id.len > 128)
        return error.InvalidRequestId;
    try validatePayload(request.operation, request.payload_schema, request.payload);
    return parsed;
}

pub const InspectSourcePayload = struct {
    source_path: []const u8,
    schema_path: ?[]const u8 = null,
};

pub const ChangeOperation = enum { add, replace, remove };
pub const ChangeItem = struct {
    change_id: []const u8,
    path: []const u8,
    operation: ChangeOperation,
    expected_old: ?std.json.Value = null,
    proposed_value: ?std.json.Value = null,
    explanation: []const u8,
};
pub const Proposal = struct {
    proposal_id: []const u8,
    revision: u64,
    source_digest: []const u8,
    created_at: []const u8,
    producer: ?[]const u8 = null,
    items: []const ChangeItem,
};
pub const ReviewComment = struct {
    comment_id: []const u8,
    change_id: []const u8,
    body: []const u8,
};
pub const RevisionPayload = struct {
    proposal_id: []const u8,
    base_revision: u64,
    source_digest: []const u8,
    allowed_change_item_ids: []const []const u8,
    comments: []const ReviewComment,
    proposal: Proposal,
    source_context: ?std.json.Value = null,
    authorized_secrets: ?std.json.Value = null,
};
pub const RevisionValidationPayload = struct {
    active_proposal: Proposal,
    candidate_proposal: Proposal,
    allowed_change_item_ids: []const []const u8,
    resolved_comment_ids: []const []const u8,
};
pub const FinalAssemblyPayload = struct {
    source_path: []const u8,
    source_digest: []const u8,
    proposal: Proposal,
    decisions: std.json.Value,
    comments: []const ReviewComment,
    external_checks: []const std.json.Value,
};
pub const ApplyPayload = struct {
    source_path: []const u8,
    source_digest: []const u8,
    final_change_digest: []const u8,
    confirmation_nonce: []const u8,
};

fn validatePayload(operation: Operation, schema: []const u8, payload: std.json.Value) !void {
    const expected = switch (operation) {
        .protocol_info => "zconfig.empty/1",
        .inspect_source => "zconfig.inspection/1",
        .validate_proposal => "zconfig.proposal/1",
        .validate_revision => "zconfig.revision/1",
        .assemble_final => "zconfig.final-change/1",
        .apply_final => "zconfig.apply/1",
    };
    if (!std.mem.eql(u8, schema, expected)) return error.InvalidPayloadSchema;
    if (operation == .protocol_info) {
        const object = switch (payload) {
            .object => |value| value,
            else => return error.InvalidOperationPayload,
        };
        if (object.count() != 0) return error.InvalidOperationPayload;
    }
}

pub fn sanitizeOutput(allocator: std.mem.Allocator, input: []const u8, secrets: []const []const u8) ![]u8 {
    var current = try allocator.dupe(u8, input);
    errdefer allocator.free(current);
    for (secrets) |secret| {
        if (secret.len == 0) continue;
        const replaced = try std.mem.replaceOwned(u8, allocator, current, secret, "[REDACTED]");
        allocator.free(current);
        current = replaced;
    }
    return current;
}

pub fn capabilityInfo() ProtocolInfo {
    return .{
        .protocol_versions = &.{protocol_version},
        .operations = &.{ "protocol_info", "inspect_source", "validate_proposal", "validate_revision", "assemble_final", "apply_final" },
        .schemas = &.{ "zconfig.empty/1", "zconfig.protocol-info/1", "zconfig.inspection/1", "zconfig.proposal/1", "zconfig.revision/1", "zconfig.final-change/1", "zconfig.apply/1" },
        .limits = .{ .message_bytes = max_message_bytes },
    };
}

pub fn writeResponse(request: Request, writer: *std.Io.Writer) !void {
    if (request.operation == .protocol_info) {
        const response = .{
            .protocol_version = protocol_version,
            .request_id = request.request_id,
            .ok = true,
            .result_schema = "zconfig.protocol-info/1",
            .result = capabilityInfo(),
        };
        return std.json.Stringify.value(response, .{}, writer);
    }
    const response = Failure{
        .request_id = request.request_id,
        .@"error" = .{
            .code = "protocol.operation_unavailable",
            .message = "Operation is recognized but not implemented.",
            .retryable = false,
        },
    };
    try std.json.Stringify.value(response, .{}, writer);
}

File: tui/internal/protocol/protocol.go
package protocol

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"regexp"
	"sort"
	"strings"
	"unicode/utf8"
)

const (
	Version         = "1.0"
	MaxMessageBytes = 16 << 20
)

type Operation string

const (
	OperationProtocolInfo     Operation = "protocol_info"
	OperationInspectSource    Operation = "inspect_source"
	OperationValidateProposal Operation = "validate_proposal"
	OperationValidateRevision Operation = "validate_revision"
	OperationAssembleFinal    Operation = "assemble_final"
	OperationApplyFinal       Operation = "apply_final"
	OperationReviseProposal   Operation = "revise_proposal"
)

type Request struct {
	ProtocolVersion string          `json:"protocol_version"`
	RequestID       string          `json:"request_id"`
	Operation       Operation       `json:"operation"`
	PayloadSchema   string          `json:"payload_schema"`
	Payload         json.RawMessage `json:"payload"`
}

type Response struct {
	ProtocolVersion string          `json:"protocol_version"`
	RequestID       string          `json:"request_id"`
	OK              bool            `json:"ok"`
	ResultSchema    string          `json:"result_schema,omitempty"`
	Result          json.RawMessage `json:"result,omitempty"`
	Error           *Error          `json:"error,omitempty"`
}

type Error struct {
	Code      string        `json:"code"`
	Message   string        `json:"message"`
	Retryable bool          `json:"retryable"`
	Details   []ErrorDetail `json:"details,omitempty"`
}
type ErrorDetail struct {
	Pointer string `json:"pointer,omitempty"`
	Reason  string `json:"reason"`
}

type ProtocolInfo struct {
	ProtocolVersions []string       `json:"protocol_versions"`
	Operations       []string       `json:"operations"`
	Schemas          []string       `json:"schemas"`
	Limits           ProtocolLimits `json:"limits"`
}
type ProtocolLimits struct {
	MessageBytes int `json:"message_bytes"`
}
type InspectSource struct {
	SourcePath string `json:"source_path"`
	SchemaPath string `json:"schema_path,omitempty"`
}
type ApplyFinal struct {
	SourcePath        string             `json:"source_path"`
	SchemaPath        string             `json:"schema_path,omitempty"`
	SourceDigest      string             `json:"source_digest"`
	FinalChangeDigest string             `json:"final_change_digest"`
	ConfirmationNonce string             `json:"confirmation_nonce"`
	Proposal          json.RawMessage    `json:"proposal"`
	Decisions         map[string]string  `json:"decisions"`
	Comments          []FinalComment     `json:"comments"`
	ExternalChecks    []ValidationResult `json:"external_checks"`
}
type FinalComment struct {
	CommentID string `json:"comment_id"`
	ChangeID  string `json:"change_id"`
	Status    string `json:"status"`
}
type InspectionResult struct {
	SourceDigest string    `json:"source_digest"`
	ByteLength   int64     `json:"byte_length"`
	NodeCount    int       `json:"node_count"`
	Settings     []Setting `json:"settings"`
}
type Setting struct {
	Path        string          `json:"path"`
	Value       json.RawMessage `json:"value"`
	Sensitivity string          `json:"sensitivity"`
	Title       string          `json:"title,omitempty"`
	Description string          `json:"description,omitempty"`
}
type ValidationResult struct {
	SubjectDigest string        `json:"subject_digest"`
	Checks        []CheckResult `json:"checks"`
}
type CheckResult struct {
	Kind           string `json:"kind"`
	Status         string `json:"status"`
	Code           string `json:"code"`
	Message        string `json:"message"`
	Pointer        string `json:"pointer,omitempty"`
	EvidenceDigest string `json:"evidence_digest,omitempty"`
}
type ExternalValidatorRequest struct {
	CandidatePath   string `json:"candidate_path"`
	CandidateDigest string `json:"candidate_digest"`
	SourceDigest    string `json:"source_digest"`
}
type ExternalValidatorResult struct {
	CandidateDigest string          `json:"candidate_digest"`
	Status          string          `json:"status"`
	Code            string          `json:"code"`
	Message         string          `json:"message"`
	Details         json.RawMessage `json:"details,omitempty"`
}

type Visibility string

const (
	VisibilityPlain    Visibility = "plain"
	VisibilityRedacted Visibility = "redacted"
)

type ValueType string

const (
	ValueTypeString  ValueType = "string"
	ValueTypeNumber  ValueType = "number"
	ValueTypeInteger ValueType = "integer"
	ValueTypeBoolean ValueType = "boolean"
	ValueTypeNull    ValueType = "null"
	ValueTypeArray   ValueType = "array"
	ValueTypeObject  ValueType = "object"
)

type RedactedValue struct {
	Visibility  Visibility `json:"visibility"`
	SecretRef   string     `json:"secret_ref"`
	ValueType   ValueType  `json:"value_type"`
	Fingerprint string     `json:"fingerprint"`
}

func (v RedactedValue) Validate() error {
	if v.Visibility != VisibilityRedacted || v.SecretRef == "" {
		return errors.New("invalid redacted value")
	}
	if !regexp.MustCompile(`^hmac-sha256:[0-9a-f]{64}$`).MatchString(v.Fingerprint) {
		return errors.New("invalid redaction fingerprint")
	}
	switch v.ValueType {
	case ValueTypeString, ValueTypeNumber, ValueTypeInteger, ValueTypeBoolean, ValueTypeNull, ValueTypeArray, ValueTypeObject:
		return nil
	}
	return errors.New("invalid redacted value type")
}

func Sanitize(data []byte, secrets []string) []byte {
	out := string(data)
	ordered := append([]string(nil), secrets...)
	sort.SliceStable(ordered, func(i, j int) bool { return len(ordered[i]) > len(ordered[j]) })
	for _, secret := range ordered {
		if secret != "" {
			out = strings.ReplaceAll(out, secret, "[REDACTED]")
		}
	}
	return []byte(out)
}

func EncodeRequest(w io.Writer, request Request) error { return json.NewEncoder(w).Encode(request) }
func DecodeRequest(r io.Reader, limit int64) (Request, error) {
	var request Request
	if err := decodeStrict(r, limit, &request); err != nil {
		return request, err
	}
	if request.ProtocolVersion != Version || request.RequestID == "" || len(request.RequestID) > 128 || !validOperation(request.Operation) || !schemaName(request.PayloadSchema) || len(request.Payload) == 0 {
		return request, errors.New("invalid request envelope")
	}
	return request, nil
}
func DecodeResponse(r io.Reader, limit int64, requestID string) (Response, error) {
	var response Response
	if err := decodeStrict(r, limit, &response); err != nil {
		return response, err
	}
	if response.ProtocolVersion != Version || response.RequestID != requestID {
		return response, errors.New("invalid response envelope")
	}
	if response.OK {
		if response.ResultSchema == "" || len(response.Result) == 0 || response.Error != nil {
			return response, errors.New("invalid success response")
		}
	} else if response.Error == nil || response.ResultSchema != "" || len(response.Result) != 0 {
		return response, errors.New("invalid error response")
	}
	return response, nil
}
func DecodePayload(raw json.RawMessage, dst any) error {
	return decodeStrict(bytes.NewReader(raw), MaxMessageBytes, dst)
}

func decodeStrict(r io.Reader, limit int64, dst any) error {
	if limit <= 0 {
		return errors.New("invalid size limit")
	}
	data, err := io.ReadAll(io.LimitReader(r, limit+1))
	if err != nil {
		return err
	}
	if int64(len(data)) > limit {
		return errors.New("message exceeds size limit")
	}
	if !utf8.Valid(data) {
		return errors.New("invalid UTF-8")
	}
	if err := rejectDuplicateNames(data); err != nil {
		return err
	}
	dec := json.NewDecoder(bytes.NewReader(data))
	dec.DisallowUnknownFields()
	dec.UseNumber()
	if err := dec.Decode(dst); err != nil {
		return err
	}
	var extra any
	if err := dec.Decode(&extra); err != io.EOF {
		if err == nil {
			return errors.New("extra JSON value")
		}
		return err
	}
	return nil
}
func rejectDuplicateNames(data []byte) error {
	dec := json.NewDecoder(bytes.NewReader(data))
	dec.UseNumber()
	var walk func() error
	walk = func() error {
		tok, err := dec.Token()
		if err != nil {
			return err
		}
		d, ok := tok.(json.Delim)
		if !ok {
			return nil
		}
		switch d {
		case '{':
			seen := map[string]bool{}
			for dec.More() {
				k, err := dec.Token()
				if err != nil {
					return err
				}
				name, ok := k.(string)
				if !ok {
					return errors.New("object name is not a string")
				}
				if seen[name] {
					return fmt.Errorf("duplicate object name %q", name)
				}
				seen[name] = true
				if err := walk(); err != nil {
					return err
				}
			}
			_, err = dec.Token()
			return err
		case '[':
			for dec.More() {
				if err := walk(); err != nil {
					return err
				}
			}
			_, err = dec.Token()
			return err
		}
		return errors.New("invalid delimiter")
	}
	if err := walk(); err != nil {
		return err
	}
	if _, err := dec.Token(); err != io.EOF {
		return errors.New("extra JSON value")
	}
	return nil
}
func validOperation(op Operation) bool {
	switch op {
	case OperationProtocolInfo, OperationInspectSource, OperationValidateProposal, OperationValidateRevision, OperationAssembleFinal, OperationApplyFinal, OperationReviseProposal:
		return true
	}
	return false
}

var schemaRE = regexp.MustCompile(`^zconfig\.[a-z0-9-]+/[1-9][0-9]*$`)

func schemaName(s string) bool { return schemaRE.MatchString(s) }
